### DISCUSSION TITLE
Updating a field relative to its current value (in one request)

### DISCUSSION THREAD

**Discussion by colemahatmccreary** [directus/directus#15313] [Feature Requests]
Couldn't find any references to this, so figured i'd make one: Directus could benefit from the ability to update a field _based on its current value_.

This would create in a world of transactional controls with a single API request, preventing the need for sequential API calls (read then update = stale data 🤮) or writing pure SQL/going around Directus entirely. 

I imagine the team has thoughts on how to best implement this, but consider this approach using a simplified ecom example:
```
item (inventory skus)
- sku
- name
- quantity_available (INT with a "must be >= 0" constraint)

order (customer orders)
- id
- items (m2m w/ item via order_item)

order_item (ordered line items)
- id
- order_id (m2o/fk)
- item_sku (m2o/fk)
- quantity_ordered (INT)
```

then some cart-driven logic to build an order request payload like
```js
// sample cart item data
let cart_items = [
  {
    sku: 'EXAMPLE_PRODUCT_1'
    quantity_in_cart: 2
  },
  {
    sku: 'EXAMPLE_PRODUCT_2'
    quantity_in_cart: 1
  }
]

// execute the directus request (using SDK syntax for simplicity)

let created_order = await directus.items('orders').createOne({
  // create each order.item
  items: cart_items.map((cart_item) => ({

    // how many of this sku in the order
    quantity_ordered: cart_item.quantity_in_cart,

    // reference the sku while updating its quantity
    item_sku: {
      sku: cart_item.sku,
      // encapsulated logic for modifying the related item quantity
      quantity_available: { 
        _relativeUpdate: (current_field_value) => current_field_value - cart_item.quantity_in_cart
      },
    }
  }))
})

```

This imagines a quantity_available **integer** field provided with an **object**, including a **_relativeUpdate** key. _relativeUpdate would accept a function, with the first argument being a placeholder for the field's current data and the return value being the updated value. 

The logic encapsulated in the function would be merged/converted into the query and any execution failures from there (due to field constraints, return type mismatch, etc) would trigger a rollback as usual. Simple math and string formatting/concatenation would be the baseline for this but I imagine folks will want to cover other cases as well.

So returning to the ecommerce example, the >= 0 check constraint on **item.quantity_availablle** would prevent orders which would result in a negative quantity.

Thoughts?

Seems adjacent to this dynamic relational filtering idea ([this issue](https://github.com/directus/directus/discussions/2993)), and I definitely like the ergonomics of the ```$VALUE()``` syntax, so maybe that can be extended.

Edit: I definitely understand accepting/evaluating a user provided function presents a significant vulnerability, but I imagine the new siloed execution environment from Flows' script operation could help solve this?

**Comment by rijkvanzanten** [comment]
I think I second your sentiment here: _Seems adjacent to this dynamic relational filtering idea (https://github.com/directus/directus/discussions/2993), and I definitely like the ergonomics of the $VALUE() syntax, so maybe that can be extended._ It'd be very useful if you could do some "light modifications" on the $VALUE() use, or maybe combine it with some other yet-to-be-created function types, like `$SUM`, so you'd get something like:

```js
await directus.items('orders').createOne({
  items: [
    {
      item_sku: {
        sku: 123,
        quantity_available: "$SUM($VALUE(quantity_available) - 1)"
      }
    }
  ]
});
```

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
